### DISCUSSION TITLE
Subject: Inquiry about Advanced Vehicle Communication and Warning System for Road Safety

### DISCUSSION THREAD

**Discussion by BenYaho** [openTCS/opentcs#98] [Q&A] [answered]
HI,

I am inquiring whether TCS has a feature that allows real vehicles to communicate and send advance warnings when they are about to meet on the road. Specifically, I would like to know if there is a "flag" or similar mechanism to prompt vehicles to stop or yield before their meeting point. If this feature is available, I am also interested in understanding how it is implemented to achieve this purpose.

Thank you for your attention and response.

**Comment by swltr** [comment]
@BenYaho There is no mechanism to warn vehicles about future collisions. The traffic management in openTCS (primarily in the form of `Scheduler` interface implementations) ensures that resources (i.e. the points and paths in the plant model) are allocated only for a single vehicle at a time. Provided that the distance between the points in your plant model is equal to or greather than the length of a vehicle or you make use of blocks, a vehicle (driver) should never get a movement command to a point that cannot be reached safely.

**Reply by BenYaho** [reply]
HI @swltr ,
The previous response mentioned resource management and control based on the distance between points and vehicle length. I'd like to inquire whether this control is based on the plant model or the layout model. Additionally, are these functionalities currently implemented or do they require further development? Your clarification would be greatly appreciated.

**Reply by swltr** [reply]
@BenYaho I'm not sure what your definition of "plant model" vs "layout model" is. However, openTCS ensures that the resources required for each step on a vehicle's route - i.e. a point and the path leading to it, and any blocks they might belong to - are allocated (exclusively) to this single vehicle before a movement command is sent to the vehicle. Thus, if the real, physical distance between points in your plant model is at least as great as the real, physical length of the vehicle, movement of vehicles between allocated points is always safe. The distance between the points in your openTCS model does not matter here.

In addition to the above, since openTCS 5.9, the resources allocated for a vehicle are only made available for other vehicles once the occupying vehicle has safely left the point/path. For this, the vehicle's length information in the plant model is used.

**Reply by BenYaho** [reply]
HI @swltr,

I want to know whether the comparison between points is to use Point Xpoistion, Point Y position or X-position(model) y-position(model) for comparison


![image](https://github.com/openTCS/opentcs/assets/128113709/32908395-62ad-4568-8876-ce8ff190e521)

**Reply by swltr** [reply, marked as answer]
It depends. If you're refering to

> Provided that the distance between the points in your plant model is equal to or greather than the length of a vehicle or you make use of blocks, a vehicle (driver) should never get a movement command to a point that cannot be reached safely.

or

> Thus, if the real, physical distance between points in your plant model is at least as great as the real, physical length of the vehicle, movement of vehicles between allocated points is always safe. The distance between the points in your openTCS model does not matter here.

then, as mentioned, the distance between the points (and thus their coordinates) in the openTCS plant model do not matter at all. I was only refering to the physical distances here.

If you're refering to

> In addition to the above, since openTCS 5.9, the resources allocated for a vehicle are only made available for other vehicles once the occupying vehicle has safely left the point/path. For this, the vehicle's length information in the plant model is used.

then, again, the coordinates in the plant model do not matter. Instead, the lengths of the paths in the plant model are used for this comparison. (You can enter a length for each path in the model editor.)

The difference between model and layout coordinates is explained in the [User's Guide](https://opentcs.org/docs/5.9/user/opentcs-users-guide.html#_layout_coordinates_vs_model_coordinates), in case that helps.
